### ISSUE TITLE
Madvoc 3->4: Body is empty for `req.getReader()` as well ass `@In @Scope(BODY)`

### ISSUE THREAD

**Issue by bentolor** [oblac/jodd#572]
## Current behavior

Sorry, to nag again :-/

After porting an Madvoc application from v3.x to 4.x it cannot read any PUT Body content. The originally, ported form was:

````
       @Action("/list/{id}") @PUT
	public JsonResult updateList(@In("id") String id) {
            return GroceryService.get().updateGroceryList(id,  request.getReader());
	}
````

I modified it to this form which does not help:

````
        @Action("/list/{id}") @PUT
	public JsonResult updateList(@In("id") String id, @In @Scope(BODY) String listJson) {
            return GroceryService.get().updateGroceryList(id,  new StringReader(listJson));
	}
````

I'm a little bit lost and hope I do not miss anything. Tested it with two web servers to sort out the issue thre.

## Expected behavior

`request.getReader()` and/or ` @In @Scope(BODY) String listJson` should contain the actual request body text. 

## Steps to Reproduce the Problem

````
git clone --depth=3 --branch=master https://github.com/bentolor/microframeworks-showcase.git showcase
cd showcase
./gradlew :jodd:run
````
Then go to http://localhost:8080/, Click "Lists" in the left section, "Edit" in the List, Remove an item, press Save.

OR

Open Project with IDEA IntelliJ and Run "Jodd: Run Example".

OR

```
./gradlew :jodd:jettyRun
````

to use Jetty instead of Undertow.

Any ideas what I'm missing?

**Comment by igr** [comment]
Ok, this one I need to check, it seems your using it correctly the `@Scope`. Sorry for the trouble, will be back soon.

**Comment by igr** [comment]
ok, @bentolor here are the results! Sorry for waiting...

1. First, the `String` type issue. Currently how *Madvoc* idea behind the `BODY` scope is to _parse_ the body into the target type. Since target type is `String` in your case, the parsing fails. So that explains the issue you are having.
2. I can fix that easily - when the target type is `String`, do not parse. The only edge case is when the body is really a JSON string like `"string\u123"` - that supports escaping and other things. In this edge case the string would not be parsed. I do not know how to distinguish this special case.
3. The idea behind `BODY` is to use the destination type right away - in your case, it would be the `GroceryList`. No need to parse for yourself! However, parsing fails at the moment since I didn't add the `UUID` JSON converter. I will add that immediately.

But wait! You can make it work now, like this:

    public JsonResult updateList(
        @In("id") String id,
        @In @Scope(SERVLET) String requestBody)

why? because `SERVLET` scope contains set of reserved names that can be used for injecting various things. I have some documentation [here](https://jodd.org/madvoc/injection.html#in) but not a complete list of reserved names.

So to summarize my action points:

+ [ ] add UUID Json type converter
+ [ ] update doc to list all reserved names
+ [ ] should `BODY` scope only deal with JSON _parsing_? If not, how to distinguish between JSON strings and rawbody?
+ [ ] Should I add simple placeholders for all reserved names instead? For example:

    @In @Scope(SERVLET) BodySupplier body

wdyt?

**Comment by bentolor** [comment]
Hi @igr 

thanks for having a look on my example so quickly. 

I was aware that I can later on directly use a typed argument. But in the first step I wanted to stick with the plain text to be in control of the deserialization/serialization i.e. to configure the expected `java.util.Date` format. 

As the ID is derived from the URL I'd also expect that Madvoc does not touch the Body unless advised and therefore `request.getReader()` still continues to work.

> should BODY scope only deal with JSON parsing?

Your comment lets me a little bit in doubt: Is `Scope.BODY` directly tied to a JSON Datatype or specific data serialization mechanism? What If the Body i.e. contains XML-Content? 

The whole statement comes a little puzzling to me, because I understood the `Scope` directive as a way to tell Jodd in **which scope to look** for or write to desired values, but **not in which data format** it will be stored there.  

It's clear to me that Madvoc needs some form of data type conversion for `@In` and `@Out`. But I'm currently wondering where I can control it in general. And even more: if I can have different conversion rules for different areas (i.e. lets say I have two `@InOut java.util.Date` targets, but in the one case I expect `2018-01-01` and in the other place a full ISO Datetime `2018-01-01 11:12:33`.

So here are a few counter questions

1. How can I configure the Data type conversion applied to  `@In` and `@Out` in genereal? (ie. XML or aspects of JSON format)
2. Can I override the applied conversion for specific places (as described above)?
3. Why did `request.getReader()` stop working though all other values are only from the URL? I'd expect it should continue to work.

> Should I add simple placeholders for all reserved names instead? 

To be honest, the whole concept of reserved name i a little new to me and I miss a section named "Reserved/Predefined injection names" or alike explaining in a short table what I get.

If I understand you correctly, depending on the scope I can use different, predefined names to access preexisting values in that Scope like:

```
 @In("request") @Scope(ScopeType.SERVLET)
 HttpServletRequest r;
 @In("body") @Scope(ScopeType.REQUEST)
 String b;
```

Am I right?

If yes, I'd vote for some form of `XXXX.SERVLET_REQUEST` and  `XXXX.REQUEST_BODY` or alike. Not sure what `XXX` should be. `InjectionParameters.SERVLET_REQUEST` ?

**Comment by igr** [comment]
@bentolor you are right! Let's make it beautiful :)

Let's go over few topics we have here :)

## [1] Existing injections and reserved names

Since `REQUEST` scope was meant for getting only the request query params and attributes, we have the `SERVLET` scope, that is used to retrieve many different things. Your last examples should be written like this (I'm writing this just for the sake of documentation):

    @In("request") @Scope(SERVLET)
    HttpServletRequest r;
    @In("requestBody") @Scope(SERVLET)
    String b;

The full list of reserved names is listed in `ServletContextScopeInjector` (yeah, I just forgot we have them :):

```
public static final String REQUEST_NAME = "request";
public static final String SESSION_NAME = "session";
public static final String CONTEXT_NAME = "context";
public static final String REQUEST_MAP = "requestMap";
public static final String REQUEST_PARAM_MAP = "requestParamMap";
public static final String REQUEST_BODY = "requestBody";
public static final String SESSION_MAP = "sessionMap";
public static final String CONTEXT_MAP = "contextMap";
public static final String CSRF_NAME = "csrfTokenValid";
```

I will change these names and remove the `NAME` suffix.

`SERVLET` scope is aware of the _types_, too - meaning, if you put a `ServletReuest` field you don't need to use above reserved name:

```
@In @Scope(SERVLET) HttpServletRequest request;
```

I believe this is a nice feature; just need to document it better :) Please let me know how this existing feature looks; personally, I find this very convenient. The only thing I am not happy with is where to put the definition of the names, currently, they are in `ServletContextScopeInjector` (where they are actually used); but this class is kind of 'internal' class, and users really do not need to be aware of it - for that reason it makes more sense to define it in the `ScopeType.SERVLET` as that class defines, as you correctly noticed, the _source_ where the data comes from.

Wdyt?

## [1.1] Replace reserved names with types

Instead of having reserved names (like above), we can simply replace them with the custom types. This way we don't have to wonder where to define them :)))

It would look like this:

```
@In
RequestBody b;
...
b.get(); // to retrieve the body.

```

Of course, each class would have `NAME` static fields, so the following would be ok:

```
@In(RequestBody.NAME)
RequestBody b;
```


wdyt?

## [2] Simple conversion and mappings

Let me quickly answer how to make simple conversions for input/output parameters. `Madvoc` support annotating the setter methods:

```
Date date;
@In
void setDate(String dateString) {
    this.date = MyUtil.convert(dateString);
}
```

and/or you can simply annotate field, but if setter and/or getter exist, `Madvoc` will use them:

```
@In String foo;
void setFoo(String myFoo) {
    this.foo = MyUtil.convert(myFoo);
}
```

This is not so a feature of `Madvoc` but of our `BeanUtil` super bean class :) All _type conversions_ are also set in `BeanUtil`; for example, I've added `UUID` converter there.

Users can use this feature to perform JSON/XML conversion as well:

```
GroceryList groceryList;
@In(requestBody) @Scope(SERVLET)
void setGroceries(String body) {
    this.groceryList = JSONParser.create().parse(body);
}
```

Not saying this is _the_ solution, but I find this fine if it is used in few places.

## [3] Finally, the _source_ vs _format_ :)

Excellent point! I agree. The scope should define _the source_.

How we can define the format? So far we are using `@Action` annotation to specify different behavior. For example:

```
@Action public void foo() {...}
@RestAction public void bar() {...}
```

These two actions have different behaviour - the mapping can be different, the results processing can be different... If we go in this direction, you would have e.g. `@XMLAction` that does different parsing.

But this approach makes things less 'visible', as the `@In` annotation may be set outside of the method - and now the same annotated field may be set differently depending of the action annotation, e.g.:

```
@In String body

@Action foo1()
@RawAction foo2()
```

For this reason I would not like this approach, but would like to specify the formatting on the field itself.

_What would be the best way to do so?_

+ `@In(formater=MyFormatter.class)` - nah, dont like so much
+ `@DataType` - i like this better! to have an annotation that actually defines _which_ type of data it is, and then you can register your own data formatter. e.g. `@DataType(JSON)`, `@DataType(XML)`, `@DataType(RAW)`. The annotation argument is a string so it is easy to extend. Byt I would like better name then `Datatype` - maybe `ContentType`, but this is not a real content type (exampe with two different date types).
+ ?

So, let's define this! I would be happy to implement, as I have it on todo for some time.


(SORRY FOR LONG POST:)

**Comment by bentolor** [comment]
Hi @igr 

again – your time and positive energy, which is very appreciated!

I'm trying to be short:

#### [1] Existing injections and reserved names

* Better documentation regarding reserved names would be a great improvement! Honestly I skipped "Context and Servlet scope" and did not understand its intention. It was too confusing for me. I would expect it more like

> Injectable Servlet and Request properties
> The Madvoc SERVLET scope allows you to access and use the following values which can be accessed by the following injection names. 
>
> || Name || Constant |Type || Value ||
> | request |  XXXX.REQUEST  | javax.servlet.HttpServletRequest | The … |
>
> Hint: For HttpServletRequest, HttpSession, ServletContext you can ommit the name. Madvoc will autodetec on the type.
> [existing example expanded by a) a Method example to access the request body and b)  maybe a example how to access URL request parameters] 

I love `@In @Scope(SERVLET) HttpServletRequest request;` !

The placement & naming of the injection names should clearly allow the user to incept that it's linked to the `SERVLET` context. Honestly – I'd probably always fall into the expectation that i..e Request Parameters and Body are to be found within the `REQUEST` scope. And after reading the Javadoc of ScopeType even more. 

#### 1.1 Replace reserved names with types

Hmmm – Pro: Typesafe. Con: Type proliferation; Maybe unexpected to use custom types to access well-defined attributes for the servlet scope. 

Would probably stick to names: Can be transparently expanded. And eventually referring injection values by a String-based named is IMHO the foundation mechanism it alway boils down to. Inferring the injection name via types is only a nice convenience, not a good design foundation.

#### Type conversion during injection
If I understand you correctly, there is currently no design infrastructure for implicit type conversion beyond the defaults provided by `BeanUtil`. This might cause issues i.e. as in my current example. 

As far as I understand this will not work for nested properties i.e. when I receive `GroceryList` and need special handling for `GroceryList.date`. 

Given the "microframework" scope I think your way of using additional setter Methods as Injection Target is a viable way to keep things simple. 

> #### Type conversion
> Madvoc uses BeanUtil to convert the between the source and target types during the injection. Default converters are available for most Java platform types like …,…,.….  
> If you need to do custom type conversion you can add an additonal setter method and  redirect the injection point via `@In`
> ````
> Date date;
> void setDate(Date date) {     this.date = date; }
> Date getDate() {     return this.date; }
> @In void setDate(String dateString) {     this.date = MyUtil.convert(dateString); }
> ````

I'm currently not sure if the upper would work (due to the two setters) and if this would also work for nested properties (like in my case the root injectable being `GroceryList`)

* [ ] Add a section "Type conversion" in Madvoc/Injection. 
* [ ] Fix 404 Link "Jodds type converter utility" in https://jodd.org/beanutil/
* [ ] If `BeanUtil` supports adding custom type conversion I'm currently not aware of it.



#### [3] Data Source vs. Data format

I threw a  look into `RestAction` and then `RestActionConfig` to be able to understand what you are talking about. There I found this block

````
		setNamingStrategy(RestActionNamingStrategy.class);
		setActionResult(JsonActionResult.class);
````

"ActionResult" or "ResultType" gives me an immediate picture about what these statements are talking about. I find DataType a little troublesome as it is to generic about which data type we are talking.

Need to think on this for a good proposal. Will come back later.

**Comment by igr** [comment]
@bentolor thank you for your time, this is a great thread :)

## [1] Existing injections and reserved names

> Better documentation regarding reserved names would be a great improvement! Honestly I skipped "Context and Servlet scope" and did not understand its intention. It was too confusing for me. I would expect it more like

True, true! Awesome example, filed an issue here: https://github.com/oblac/jodd-site/issues/18 Will do. Sorry for that :(

> Honestly – I'd probably always fall into the expectation that i.e. Request Parameters and Body are to be found within the REQUEST scope.

I understand (and agree). Just have no idea how to make difference between the String body from any parameter/attribute. The only idea I have is to use custom types, e.g.:

```
@In @Scope(REQUEST) RequestBody requestBody;
```

I am not sure if there is a way to distinguish between the body and parameters using the same annotation. That's why we added the `BODY` scope.

## 1.1 Replace reserved names with types

> Would probably stick to names: Can be transparently expanded. And eventually referring injection values by a String-based named is IMHO the foundation mechanism it alway boils down to. Inferring the injection name via types is only a nice convenience, not a good design foundation.

True. Moreover, this new classes would be simple value holders and is sort-of un-natural. But I would definitly move defintion of names from current class to e.g. the `Scope` class, it makes much more sense!


## [2] Type conversion during injection

> If I understand you correctly, there is currently no design infrastructure for implicit type conversion beyond the defaults provided by BeanUtil. This might cause issues i.e. as in my current example.

Not by Madvoc (except using a setter). But let's use your case as a requirement!

> As far as I understand this will not work for nested properties i.e. when I receive GroceryList and need special handling for GroceryList.date.

Correct. 

> Given the "microframework" scope I think your way of using additional setter Methods as Injection Target is a viable way to keep things simple.

True, but... I would like to give it a try to design simple way of specifying converters for single injection; that might work for you.

> I'm currently not sure if the upper would work (due to the two setters) and if this would also work for nested properties (like in my case the root injectable being GroceryList)

Would work only for first level. Two setter of the same name would not work (probably), i usually use different names.

## [3] Data Source vs. Data format

> "ActionResult" or "ResultType" gives me an immediate picture about what these statements are talking about. I find DataType a little troublesome as it is to generic about which data type we are talking.

To recap:

+ set conversion behavior in `ActionConfig` (so user can use either a custom annotation or `@ActionConfiguredBy`). Injections are managed by the same action configuration - like you said, user would understand it. However, it is not immediately clear what is injected; moreover the number of combination of different conversions and existing configs would multiply the number of different action configs (JsonRestAction, JsonAction, PlainRestAction, PlainAction...)

+ so far as I see an additional optional annotation should be introduced:

```
@In @Type(RawString) @Scope(BODY) String body;
@In @Type(Json) @Scope(BODY) Foo foo;
@In @Type(MyJson) @Scope(BODY) GroceriesList;
```
although `type` is a terrible name here. Maybe `Mapper`?

**Comment by igr** [comment]
btw, shouldn't Jodd be able to parse both:

+ `2018-01-01` 
+ `2018-01-01 11:12:33`

by default (without any additional settings)?

**Comment by igr** [comment]
We can different injections depending on the `content-type`.

+ if `json` content-type, use JsonParser
+ if `xml` content-type, use XmlParser
+ if User wants a raw content, he can enforce content type:

```
@In @ContentType(text) @Scope(Body) String body
```

+ if user wants custom json, he can use his own content type variant:

```
@In @ContentType("application/json;customList") @Scope(Body) body
```

I believe that we might leveredge the existance of the HTTP content type and have a mapper for each content type. I was planning to add this to also for the action mapping, as for now you cant map an action to a content type.

**Comment by bentolor** [comment]
> I am not sure if there is a way to distinguish between the body and parameters using the same annotation. That's why we added the BODY scope.

Reserved name `requestBody` is not an option? 

> I would like to give it a try to design simple way of specifying converters for single injection; that might work for you.

I see two variants as a user here:

1. You have some special requirements in a few places and need/want to change it with low-key effort just in-place. `Mapper` oder [`TypeAdapter` like in Gson](https://github.com/bentolor/microframeworks-showcase/blob/master/jodd/src/main/java/bentolor/grocerylist/persistence/LocalDateAdapter.java) sound like reasonable names to me.
````
@InOut @Mapper(MyCustomFooMapper.class) @Scope(BODY) Foo foo;
@In("id") @Mapper(MyUUID2.class) UUID2 uuid;
````
2. It turns out your model is a collection of oddities; therefore you want to replace the whole type conversion infrastructure.
````
@MadvocAction(mapper = GsonMapper.class)
public class ListAction {

    @Action("/list") @GET
    public JsonResult listAllLists() { …}

    @Action("/list", mapper = XmlMapper.class) @POST
    public JsonResult createList(@In @Scope(SERVLET) String requestBody) { … }
}
````

Examples only as rough idea; not polished proposal.

I'm not sure how to get this smoothly in sync with the `ActionConfig` approach. For example, how to control the type mapping here: 

````
@RestAction("{id}") @GET  public User findUser() {  return someUser;    }
````


> btw, shouldn't Jodd be able to parse both date format?

Did not test. In any case I also need to control the returned format. 

> We can different injections depending on the `content-type`

Supporting different content types at once might be an important extension. But being able to configure the serialisation/deserialization in general is precursor for that to me.

**Comment by igr** [comment]
> Reserved name requestBody is not an option?

Hm, not sure... I personally don't like exception cases: this only word would become reserved. Then someone else might need and ask for the e.g. request's content type or for http method name and for who-knows-what, and soon the list of reserved names expands.

I was trying to find a name that is invalid for parameters and attributes, but there is none.

For that reason, I would rather use different scopes. Maybe the name 'scope' is wrong, as they indicate the servlet's scope, hence the issue; it is more "the source of the injection value" (and I do not know a better word in English for that :).

> You have some special requirements in a few places and need/want to change it with low-key effort just in-place. Mapper or TypeAdapter like in Gson sound like reasonable names to me.

Like it! Good point!

> It turns out your model is a collection of oddities; therefore you want to replace the whole type conversion infrastructure.

Got it - this way we basically define a mapper for all the conversion that happens in this action class.

> Examples only as rough idea; not polished proposal.

This is what I need the most , no worries.

> Supporting different content types at once might be an important extension. But being able to configure the serialisation/deserialization in general is precursor for that to me.

Agree. Let me dive into the code and work on the rough `Mapper` proposal.

**Comment by igr** [comment]
I didn't forget this. Before adding a mapper, I had to change how Scopes works. Now, the scope is simply defined by the annotation.

This means that you can use a default `Body` scope, but also it means that you can build a `XMLBody` scope pretty easily and use that one.

Furthermore, Ive added `InjectionPoint` - place where I could store potential `Mapper` information, for each injection point. This would mean that scope could use this information to use the correct mapper.

My only question now is should we add a mapper - as it is quite easy to build custom scopes - just make a class. Maybe it is too much for little Jodd? :)

**Comment by bentolor** [comment]
I'm not sure if I'm able to follow:

* https://github.com/oblac/jodd/blob/981f0bcc2956f80a50dab0d81d712906945d539b/jodd-madvoc/src/main/java/jodd/madvoc/config/InjectionPoint.java sounds like a reasonable place to declare transformation
* But I'm now completely uncertain about how the new API design looks like either for a `XMLBody` case nor a using `InjectionPoint` to perform custom data type transformations

**Comment by igr** [comment]
@bentolor here is the idea, not yet documented as it is totally fresh.

From now on the scopes are defined in annotations by specifying the scope name. The scope is a 'thing' that actually performs injection/outjection of a value from the context into the `Target`. The scopes are singletons.

The scope name can be either class name or some random name you make it. So the whole process goes like this:

1. OPTIONALLY register manually the scope: `scopeResolver.registerScope(XmlBody.class);` or `scopeResolver.registerScope("name", XmlBody.class)`. `ScopeResolver` is part of the Madvoc components and it is easy to reach any time during the initalization.
2. Define a `String` constant with either the registered name from step 1 or just equals to new scope class name.
3. use your name in `Scope` annotation
4. Create your `MadvocScope` class.

Basically, you are doing this:

```
	@Scope(XmlBody.class)
    public void myaction()
```

I didn't want to depend on the class (due to Java limitations) but the `String`, so what you do is this:

```
    @Scope(XMLBODY)
    public void myaction()
```

where `XMLBODY` is a String constant with new scope name or full class name.

THe scope is easy to write:

```
public class XmlBodyScope implements MadvocScope {

    @PetiteInject
    you_can_inject_anything here

	@Override
	public void inject(final ActionRequest actionRequest, final Targets targets) {
		final String body = actionRequest.readRequestBody();
		if (StringUtil.isEmpty(body)) {
			return;
		}

		targets.forEachTargetAndIn(this, (target, in) -> {
			final Object value = MYXML_CONVERSION(body, in.type()); // CONVERT!

			target.writeValue(in.propertyName(), value, true);
		});
	}
```

Once again: you just write your own `MadvocScope` type that does the conversion. There is just this additional step of having the `String` name registration, instead of realying on the class.

P.S. I could remove the String names for scopes, but since Java allows only literals, you would be needed to write `.class` on every `Scope` annotation, and I simply didnt like that. You can't have a static constant of class and use that in annotation.

**Comment by igr** [comment]
With latest commit it is even easier, you can just extend default BodyScope class and overide a parsing method:

https://github.com/oblac/jodd/commit/81cdeef0538cdeaf4d57a44a0058be95f01d22e1#diff-642f7bbe1cfa64fa27c9ecaacdce1162R57

**Comment by bentolor** [comment]
@igr Currently very busy; will try to respond on weekend. Keep tight!

**Comment by bentolor** [comment]
I tried to understand the idea behind scopes. Let me see if I was able to get it straight:

1. "Scopes" are implementations which are able to inject/outject **a set of values** from a **source context** retrieved via a **fixed, shared set of methods** declared by `MadvocScope` into **one injection target**. 
1. Purely for simplification, your MadvocScope declares String constants referring to these different implementation. AFAIK this is motivated by the search for an elegant API
1. You are contemplating about using _new, custom scope_ to reflect different source formats. This is mostly motivated due to the fact, that currently "BodyScope" is bound to JSON input

### A few comments about the general design in detail
1. "1." already limits somehow the set of possible "context". Because you only have a fixed set of methods to retrieve you source context. This makes it IMHO rather unlikely that so. wants to create a new bunch of new "scopes"
1. "2." breaks a basic principle of design: The superclass (indirectly) refers its subclasses. Introducing a new subclass would require you to change the superclass on any change. I don't think this is generally a bad idea, if **we agree, that the set of Scopes are rather fixed**
1. More confusingly – if the set of options are already fixed, why – as a user – do I need to discover `MadvocScope` to find out how to use the `Scope` annotation class typically? 
1. Regarding data conversion/data mapping I think we have two layers
      1. converting **from/to the source context** into **a generic value map** and
      1. converting **one value stored within the context  value map** into **a target type very specific to the specific injection target point**

### My POV on data transformation
* The current `BodyScope` is a very unique one in the set of predefined Scopes, because in most other Scopes. **the scope itself already provides a direct way to read/access single source values**. **Only in body, the scope itself needs to "understand" a source data format**.
* My primary focus was pointed **at the data conversion required to transform (typically a string) from the source context via a way very specific for the very specific injection target field/parameter**.

### WWBC (what-would-bentolor-code)
I would try to keep it plain, simple and stupid:
* Introduce a `BodyScope` and a `JsonBodyScope`. The later is what your current `BodyScope` does. The former is simply a "one value as string : the request body text". If you already have support for XML maybe I'd add also a `XmlBodyScope`. But its not needed because of the following:
* I'd also add support, to be able to declare a custom data mapper class for the conversion of the value I receive from the context into the target type.
* And last but not least: Contemplate about, transforming `@Scope(MadvocScope.BODY)` directly into some `@Body`-Annotation

This could i..e. look like this:

    @Action("/list") @POST
    public JsonResult createList(@In @Body @Mapper(MyTypeMapper.class) MyType mySpecialType) { … }

    @Action("/list") @PUT
    public JsonResult createList(@In @JsonBody MyType receivedfromJson) { … }

    @Action("/list") @PUT
    public JsonResult createList(@InOut @Cookie String sessionId) { … }

### To stick to your original question

> My only question now is should we add a mapper - as it is quite easy to build custom scopes - just make a class. Maybe it is too much for little Jodd? :)

I think the granularity of the Scope is not appropriate, because you need to be able to control the mapping **per injection field** and **not per source context**.

> @Scope("name") vs @Scope(Name.class)

Because if you use only ` @Scope("name")` instead of ` @Scope("fully.qualified.classname")` the name itself requires Madvoc to now how to turn this string into the implementation class. So probably I'd either

1. stick to `@Scope(Name.class)` or
2. add _additional_ @Body-annotations for Convenience.

Hope this helps

**Comment by igr** [comment]
Thank you for detailed review @bentolor!

I think that is the way to go, plain and stupid is all that Jodd is about :) No kidding, that is powerful. I was not completely satisfied with my latest refactorings, but it was a step forward.

Thank, I truly appreciate it! Can I send you some Jodd laptop stickers, I have few of them around? No kidding, just email me your address :)))

**Comment by igr** [comment]
Added everything.

The difference is that `@Mapper` is now not only Madvoc but applies on any `BeanUtil`, hence it can be used in different situations as well.

**Comment by igr** [comment]
I will close this issue. This feature will come in Jodd 5.

Please let me know if you notice something weird :)

**Comment by bentolor** [comment]
> The difference is that @Mapper is now not only Madvoc but applies on any BeanUtil, hence it can be used in different situations as well.

This sounds great and absolutely appropriate!
